api/v1alpha1/proxydefaults_types.go

```diff
@@ -0,0 +1,208 @@
+	cond := in.Status.GetCondition(ConditionSynced)
+	if cond == nil {
+		return corev1.ConditionUnknown, "", ""
+	}
+	return cond.Status, cond.Reason, cond.Message
+}
+
+func (in *ProxyDefaults) SyncedConditionStatus() corev1.ConditionStatus {
+	cond := in.Status.GetCondition(ConditionSynced)
+	if cond == nil {
+		return corev1.ConditionUnknown
+	}
+	return cond.Status
+}
+
+func (in *ProxyDefaults) Name() string {
+	return in.ObjectMeta.Name
+}
+
+func (in *ProxyDefaults) SetSyncedCondition(status corev1.ConditionStatus, reason string, message string) {
+	in.Status.Conditions = Conditions{
+		{
+			Type:               ConditionSynced,
+			Status:             status,
+			LastTransitionTime: metav1.Now(),
+			Reason:             reason,
+			Message:            message,
+		},
+	}
+}
+
+func (in *ProxyDefaults) ToConsul() api.ConfigEntry {
+	consulConfig := in.convertConfig()
+	return &capi.ProxyConfigEntry{
+		Kind:        in.ConsulKind(),
+		Name:        in.Name(),
+		MeshGateway: in.Spec.MeshGateway.toConsul(),
+		Expose:      in.Spec.Expose.toConsul(),
+		Config:      consulConfig,
+	}
+}
+
+func (in *ProxyDefaults) MatchesConsul(candidate api.ConfigEntry) bool {
+	proxyDefCand, ok := candidate.(*capi.ProxyConfigEntry)
+	if !ok {
+		return false
+	}
+	proxyDefCand.Namespace = ""
+	proxyDefCand.CreateIndex = 0
+	proxyDefCand.ModifyIndex = 0
+
+	return reflect.DeepEqual(in.ToConsul().(*capi.ProxyConfigEntry), proxyDefCand)
+}
+
+func (in *ProxyDefaults) Validate() error {
+	var allErrs field.ErrorList
+	path := field.NewPath("spec")
+
+	if err := in.Spec.MeshGateway.validate(path.Child("meshGateway")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	if err := in.validateConfig(path.Child("config")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	allErrs = append(allErrs, in.Spec.Expose.validate(path.Child("expose"))...)
+	if len(allErrs) > 0 {
+		return apierrors.NewInvalid(
+			schema.GroupKind{Group: ConsulHashicorpGroup, Kind: ProxyDefaultsKubeKind},
+			in.Name(), allErrs)
+	}
+
+	return nil
+}
+
+// matchesConfig compares the values of the config on the spec and that on the
+// the consul proxy-default and returns true if they match and false otherwise
+func (in *ProxyDefaults) matchesConfig(config map[string]interface{}) bool {
+	if in.Spec.Config == nil || config == nil {
+		return in.Spec.Config == nil && config == nil
+	}
+	var inConfig map[string]interface{}
+	if err := json.Unmarshal(in.Spec.Config, &inConfig); err != nil {
+		return false
+	}
+	return reflect.DeepEqual(inConfig, config)
+}
+
+// convertConfig converts the config of type json.RawMessage which is stored
+// by the resource into type map[string]interface{} which is saved by the
+// consul API.
+func (in *ProxyDefaults) convertConfig() map[string]interface{} {
+	if in.Spec.Config == nil {
+		return nil
+	}
+	var outConfig map[string]interface{}
+	// We explicitly ignore the error returned by Unmarshall.
```

Comment:
```suggestion
	// We explicitly ignore the error returned by Unmarshall
	// because validate() ensures that if we get to here that it
	// won't return an error.
```